File: experiments/llm-reasoning-cost/generated_code/clojure/07_word_freq_mini.clj
(defn count-words [text]
  (let [words (clojure.string/split (clojure.string/lower-case text) #"\s+")]
    (reduce (fn [acc word]
              (update acc word (fnil inc 0)))
            {}
            words)))

(defn print-top [word-counts n]
  (->> word-counts
       (sort-by (fn [[_ count]] [- count (key _)]))
       (take n)
       (doseq [[word count]] (println (str "  " count " " word)))))

(defn -main []
  (let [text "the cat sat on the mat the cat sat"
        word-counts (count-words text)]
    (print-top word-counts 5)))

(-main)

File: experiments/llm-reasoning-cost/generated_code/rust/07_word_freq_mini.rs
use std::collections::HashMap;

fn count_words(text: &str) -> HashMap<String, i32> {
    let mut word_count = HashMap::new();
    for word in text.to_lowercase().split_whitespace() {
        *word_count.entry(word.to_string()).or_insert(0) += 1;
    }
    word_count
}

fn print_top(word_count: &HashMap<String, i32>, n: i32) {
    let mut sorted_counts: Vec<_> = word_count.iter().collect();
    sorted_counts.sort_by(|a, b| {
        let count_cmp = b.1.cmp(a.1);
        if count_cmp == std::cmp::Ordering::Equal {
            a.0.cmp(b.0)
        } else {
            count_cmp
        }
    });

    for (word, count) in sorted_counts.iter().take(n as usize) {
        println!("  {} {}", count, word);
    }
}

fn main() {
    let text = "the cat sat on the mat the cat sat";
    let word_count = count_words(text);
    print_top(&word_count, 5);
}
